File: pgx/internal/queryplan/mapper.go
// Copyright 2021-2026 Zenauth Ltd.
// SPDX-License-Identifier: Apache-2.0

package queryplan

import (
	"fmt"
	"strings"
)

// Relation describes how a collection- or object-valued attribute reaches another table.
//
// The correlation is always `<Table>.<TargetColumn> = <parent>.<SourceColumn>`, where the parent
// is the enclosing scope's table — the root table at the top level, or the previous hop's table
// inside a nested lambda. Getting that rebase wrong is what the corpus's `outer-attr-depth2` and
// `w2-outer-relation` actions exist to catch.
type Relation struct {
	// Field maps a scalar collection's element to its single column, e.g. `subCategoryNames`
	// where each element is the string itself rather than an object.
	Field *Entry
	// Fields maps an object collection's element fields, e.g. `tags` where `t.name` reads the
	// related table's `name` column.
	Fields map[string]Entry
	// Table is the element table — the one a lambda body reads its fields from.
	Table string
	// SourceColumn is the column on the parent scope's row that the outermost table matches.
	SourceColumn string
	// TargetColumn is the matching column on the outermost table — Table itself when Via is
	// empty, otherwise the last hop's table.
	TargetColumn string
	// Via lists intermediate tables between Table and the parent, innermost first. It is empty
	// for a direct relation and non-empty for a flattened chain such as
	// `mainCategory.subCategories`, which reaches the resource through its category table.
	Via []Hop
	// SubqueryFilter is the predicate the application itself applies when it reads Table — a
	// soft-delete flag, a tenant column, a subtype discriminator, whatever narrows the rows that
	// became the resource attributes.
	//
	// The translator reads Table bare. It is given a table name and two column names and has no
	// association metadata to consult, so nothing that narrows the application's own read reaches
	// the generated subquery, and the subquery then examines rows the application never
	// serialised — a filter that returns rows the PDP denies. Declaring the predicate here
	// restores the equality. It is optional because the translator cannot detect the omission;
	// leaving it empty emits exactly the SQL it emitted before the field existed. See "Mapping
	// hazards" in the adapter's README.
	//
	// Entries are ANDed into the subquery's correlation predicate, so the restriction narrows the
	// rows the subquery EXAMINES rather than the rows it returns. That is what keeps it right
	// under negation: `all` lowers to a false-witness EXISTS, and restricting the scan turns it
	// into "every visible row satisfies the body" instead of "every row in the table does".
	SubqueryFilter []Restriction
}

// Hop is one intermediate table in a flattened relation chain.
type Hop struct {
	// Table is the intermediate table being joined in.
	Table string
	// ChildColumn is the referencing column on the table one step further in — the element
	// table for the first hop, the previous hop's table after that.
	ChildColumn string
	// JoinColumn is the referenced column on this hop's table.
	JoinColumn string
	// SubqueryFilter restricts this hop's table the same way Relation.SubqueryFilter restricts
	// the element table. A hop is read bare too, and it is the to-ONE parent whose absence must
	// deny (#309) — a row the application's own reads hide is, for this purpose, absent.
	SubqueryFilter []Restriction
}

// RestrictOp is the comparison a Restriction applies.
type RestrictOp uint8

const (
	// RestrictEq is `column = value`. It is the zero value, so a Restriction that names only a
	// column and a value is an equality.
	RestrictEq RestrictOp = iota
	// RestrictNe is `column <> value`. It does not match NULL columns; pair it with
	// RestrictIsNull under an OR if the application's own predicate treats NULL as matching.
	RestrictNe
	// RestrictIsNull is `column IS NULL` — the usual soft-delete spelling.
	RestrictIsNull
	// RestrictIsNotNull is `column IS NOT NULL`.
	RestrictIsNotNull
	// RestrictIn is `column IN (values…)`. An empty Values list hides every row.
	RestrictIn
	// RestrictNotIn is `column NOT IN (values…)`. An empty Values list hides none.
	RestrictNotIn
)

// Restriction is one caller-declared predicate on the rows a relation subquery may see.
//
// The vocabulary is deliberately narrow — equality, inequality, null tests and membership over a
// single column — because that is what the hazards it exists for look like: a `deleted_at IS
// NULL` soft delete, a `tenant_id = $1` scope, a `kind IN (…)` discriminator. A predicate that
// does not fit is a signal that the mapping cannot faithfully reproduce the application's read,
// and the honest response is to not map that relation rather than to declare an approximation.
//
// Value and Values are read according to Op, and the pairing is not enforced at compile time. It
// does not need to be, because every way of getting it wrong FAILS CLOSED: Values alongside
// RestrictEq leaves Value nil and renders `column = NULL`, which is UNKNOWN and admits nothing,
// and Value alongside RestrictIn leaves Values empty, which folds to FALSE. A mismatched pairing
// therefore hides rows rather than exposing them. TestRestrictionMismatchFailsClosed pins it.
type Restriction struct {
	// Value is compared with Column. Ignored by RestrictIsNull and RestrictIsNotNull.
	Value any
	// Column is a column on the table this restriction applies to, unqualified.
	Column string
	// Values is the membership list for RestrictIn and RestrictNotIn.
	Values []any
	// Op is the comparison. The zero value is RestrictEq.
	Op RestrictOp
}

// ValueType marks an attribute whose stored representation needs special handling.
type ValueType uint8

const (
	// ValueDefault is an ordinary column.
	ValueDefault ValueType = iota
	// ValueTimestamp marks a column stored as a temporal type, so a `timestamp()` conversion
	// around it is a no-op rather than an unsupported shape.
	ValueTimestamp
	// ValueBool marks a column stored as a boolean. It exists so `string()` over one can fail
	// closed: SQLite and MySQL have no boolean type and render the stored 1/0 as "1", while CEL
	// and PostgreSQL render "true", and nothing in the plan says which type a column holds
	// (cerbos/query-plan-adapters#376).
	ValueBool
	// ValueString marks a column stored as text. It exists so CEL's `+` between two columns can
	// be resolved to concatenation: the operator is overloaded on strings and the plan carries no
	// operand types, so without a declaration the shape fails closed rather than emitting a
	// numeric `+` that MySQL silently answers with 0 (cerbos/query-plan-adapters#391).
	ValueString
)

// NullConvention declares, for one attribute, that its column can be SQL NULL and how the caller
// represents that NULL in the attributes it sends to check().
//
// It is per attribute rather than per call because one policy suite can legitimately mix the two
// conventions — the same column can be mapped twice, sent as an explicit null under one attribute
// name and omitted under another. Options.NullRepresentation cannot express that, which is what
// made https://github.com/cerbos/query-plan-adapters/issues/308 unfixable with the call-level
// option alone.
//
// The zero value declares nothing, which is why this is a separate type from NullRepresentation
// rather than a reuse of it: NullExplicit is NullRepresentation's zero value, and an Entry that
// declared a convention merely by existing would change the SQL emitted for every mapped column.
type NullConvention uint8

const (
	// NullConventionUnset declares nothing. The column is treated as NOT NULL when rendering a
	// comparison — the historical translation — and Options.NullRepresentation still governs
	// whether a null comparison operand is rejected.
	NullConventionUnset NullConvention = iota
	// NullConventionExplicit means a NULL column is sent as an explicit null attribute, so CEL
	// holds a null VALUE: `null != "x"` is TRUE and `null == "x"` is FALSE, both definite. SQL's
	// UNKNOWN excludes the row under BOTH polarities, so the equality family (eq, ne, in) is
	// rendered so it can never be UNKNOWN. Ordering and string operators are left alone: they
	// raise a no-overload error on a null receiver in CEL, which denies exactly as UNKNOWN does.
	NullConventionExplicit
	// NullConventionOmitted means a NULL column sends no attribute, so CEL raises a
	// missing-attribute error and check() denies. UNKNOWN already excludes the row under both
	// polarities, so the rendering is unchanged; what the declaration adds is the same
	// null-operand rejection Options.NullRepresentation = NullOmitted performs, scoped to this
	// attribute.
	NullConventionOmitted
)

// Entry is what a plan variable resolves to.
type Entry struct {
	// Relation is set when the attribute reaches another table, and Column is empty.
	Relation *Relation
	// ScalarRelation reads Column from another table through a to-ONE hop, as a correlated
	// scalar subquery — `R.attr.parent.aString`, where `parent` is a joined row rather than a
	// column on this one. It reuses Relation's vocabulary (Table, SourceColumn, TargetColumn,
	// Via, SubqueryFilter) because the chain is described identically; only the projection
	// differs, and Via carries the intermediate levels of a multi-level chain.
	//
	// It is a separate field from Relation rather than "Relation plus a Column" so the to-ONE
	// claim is explicit. The translator cannot check it: nothing in a table name says the
	// correlation matches at most one row, and against a to-MANY relation the database would
	// either pick an arbitrary row or fail at runtime depending on dialect. Declaring it here is
	// the caller asserting the uniqueness their schema enforces.
	ScalarRelation *Relation
	// Column names a column on the row this entry is read from — the enclosing scope's row, or
	// ScalarRelation's element row when that is set.
	Column string
	// Qualifier is the table or alias the column is read from. Callers normally leave it empty
	// and let the translator fill it in — the resource table at the top level, a collection
	// element's alias inside a lambda. Set it to read a column from somewhere else.
	Qualifier string
	// ValueType marks a column whose stored representation needs special handling.
	ValueType ValueType
	// NullConvention declares this column's NULL representation. See NullConvention.
	NullConvention NullConvention
}

// Mapper resolves a plan variable — the full reference as the planner emits it, e.g.
// `request.resource.attr.aString` — to the storage it lives in.
//
// Resolution is deliberately fail-closed: an unmapped reference is an error, never a guessed
// column name. Snake-casing an attribute automatically (which the original Go helpers did) turns
// a policy typo into a silent SQL error or, worse, a match against the wrong column.
type Mapper interface {
	Resolve(reference string) (Entry, bool)
}

// MapperFunc adapts a function to the Mapper interface.
type MapperFunc func(reference string) (Entry, bool)

// Resolve implements Mapper.
func (f MapperFunc) Resolve(reference string) (Entry, bool) { return f(reference) }

// MapperMap is the common case: a static table of references to entries.
type MapperMap map[string]Entry

// Resolve implements Mapper.
func (m MapperMap) Resolve(reference string) (Entry, bool) {
	e, ok := m[reference]
	return e, ok
}

// rootMapper stamps the resource table onto every entry the caller's mapper returns.
//
// Without it, an entry resolved from inside a lambda body would inherit whatever scope happened
// to be current, so `R.attr.aBool` referenced two lambdas deep would read the innermost element's
// alias instead of the resource row. Every Entry therefore carries the qualifier of the row it is
// read from — the resource table here, the element alias in scopedMapper — and the translator
// never has to guess from context.
type rootMapper struct {
	parent Mapper
	table  string
}

func (r rootMapper) Resolve(reference string) (Entry, bool) {
	entry, ok := r.parent.Resolve(reference)
	if !ok {
		return Entry{}, false
	}
	if entry.Qualifier == "" {
		entry.Qualifier = r.table
	}
	return entry, true
}

// scopedMapper resolves lambda-local references (`t.name`) against the element fields of the
// collection currently being iterated, falling back to the enclosing mapper for everything else —
// which is how an outer resource attribute stays readable from inside a nested lambda body.
type scopedMapper struct {
	parent   Mapper
	relation *Relation
	variable string
	alias    string
}

func (s scopedMapper) Resolve(reference string) (Entry, bool) {
	if reference == s.variable {
		// A bare reference to the iteration variable: only meaningful for a scalar collection,
		// where the element *is* the column.
		if s.relation.Field != nil {
			e := *s.relation.Field
			e.Qualifier = s.alias
			return e, true
		}
		return Entry{}, false
	}

	if field, ok := strings.CutPrefix(reference, s.variable+"."); ok {
		e, ok := s.relation.Fields[field]
		if !ok {
			return Entry{}, false
		}
		e.Qualifier = s.alias
		return e, true
	}

	return s.parent.Resolve(reference)
}

// requireRelation resolves a reference that must name a collection.
func requireRelation(m Mapper, reference string) (Entry, error) {
	entry, ok := m.Resolve(reference)
	if !ok {
		return Entry{}, fmt.Errorf("no mapping for collection attribute %q", reference)
	}
	if entry.Relation == nil {
		return Entry{}, fmt.Errorf(
			"attribute %q is mapped to a column but is used as a collection; map it as a relation",
			reference,
		)
	}
	return entry, nil
}

// guardedMapper rejects a caller-supplied qualifier that collides with the prefix used for
// generated subquery aliases.
//
// RootTable is checked once up front, but an Entry may name its own qualifier, and a colliding one
// would let an inner subquery alias shadow it — the correlation would compare the subquery's row
// against itself and the filter would match rows the PDP denies, with no SQL error to notice.
type guardedMapper struct {
	parent Mapper
}

func (g guardedMapper) Resolve(reference string) (Entry, bool) {
	entry, ok := g.parent.Resolve(reference)
	if !ok || strings.HasPrefix(entry.Qualifier, aliasPrefix) {
		return Entry{}, false
	}
	return entry, true
}

File: drizzle/src/mapper.ts
import type { PlanExpressionOperand } from "@cerbos/core";
import { sql } from "drizzle-orm";
import type { AnyColumn, SQL } from "drizzle-orm";

import { isNameOperand } from "./operands";
import { SCOPED_RELATION } from "./types";
import type {
  BaseMapperEntry,
  Mapper,
  MapperEntry,
  MappingConfig,
  NullAttributeRepresentation,
  RelationMapping,
  RelationValue,
  ResolvedMapping,
  ScopedRelationEntry,
} from "./types";

/**
 * Resolving a CEL attribute reference through the caller's mapper: which column, transform or
 * relation it names, and which relation chain has to be joined to reach it.
 */

// -- entry shapes --------------------------------------------------------------------------------

const isScopedRelationEntry = (
  entry: MapperEntry,
): entry is ScopedRelationEntry =>
  typeof entry === "object" && entry !== null && SCOPED_RELATION in entry;

export const isMappingConfig = (entry: MapperEntry): entry is MappingConfig =>
  typeof entry === "object" &&
  entry !== null &&
  !isScopedRelationEntry(entry) &&
  ("column" in entry ||
    "transform" in entry ||
    "relation" in entry ||
    "valueType" in entry ||
    "indexable" in entry ||
    "collectionValueType" in entry);

export const isRelationValue = (
  entry: BaseMapperEntry,
): entry is RelationValue =>
  typeof entry === "object" &&
  entry !== null &&
  "kind" in entry &&
  entry.kind === "relation";

export const isColumn = (entry: BaseMapperEntry): entry is AnyColumn =>
  typeof entry === "object" &&
  entry !== null &&
  !isRelationValue(entry) &&
  !isMappingConfig(entry) &&
  typeof entry !== "function";

/** Whether an entry declares its relation a collection of scalars (`collectionValueType`). */
export const isScalarCollection = (entry: MapperEntry | undefined): boolean =>
  entry !== undefined &&
  isMappingConfig(entry) &&
  entry.collectionValueType === "scalar";

const toBaseMapperEntry = (entry: MapperEntry): BaseMapperEntry =>
  isScopedRelationEntry(entry) ? entry.resolve().mapping : entry;

const makeScopedRelationEntry = (
  resolution: ResolvedMapping,
): ScopedRelationEntry => ({
  [SCOPED_RELATION]: true,
  resolve: () => resolution,
});

/**
 * The convention declared on a mapper entry, if it declares one. An entry that declares nothing
 * inherits the call-level option.
 */
export const mappingNullRepresentation = (
  mapping: BaseMapperEntry,
): NullAttributeRepresentation | undefined =>
  isMappingConfig(mapping) ? mapping.nullAttributeRepresentation : undefined;

// -- resolution ----------------------------------------------------------------------------------

export const getMappingEntry = (
  reference: string,
  mapper: Mapper,
): MapperEntry | undefined =>
  typeof mapper === "function" ? mapper(reference) : mapper[reference];

/** Every `[prefix, suffix]` split of a dotted reference, longest prefix first. */
function* referencePrefixes(reference: string): Generator<[string, string[]]> {
  const parts = reference.split(".");
  for (let i = parts.length - 1; i > 0; i--) {
    yield [parts.slice(0, i).join("."), parts.slice(i)];
  }
}

const resolveRelationField = (
  relation: RelationMapping,
  path: string[],
  reference: string,
  accumulated: RelationMapping[],
  allowDefaultField = true,
): ResolvedMapping => {
  const relations = [...accumulated, relation];

  if (path.length === 0) {
    if (!allowDefaultField) {
      return { relations, mapping: { kind: "relation", relation } };
    }
    if (!relation.field) {
      throw new Error(
        `Relation mapping for '${reference}' does not define a default field`,
      );
    }
    return { relations, mapping: toBaseMapperEntry(relation.field) };
  }

  const [segment, ...rest] = path;
  if (segment === undefined) {
    throw new Error(
      `Invalid relation path for reference '${reference}': missing segment`,
    );
  }
  const fieldEntry = (relation.fields ?? {})[segment];

  if (fieldEntry !== undefined) {
    if (isMappingConfig(fieldEntry) && fieldEntry.relation) {
      return resolveRelationField(
        fieldEntry.relation,
        rest,
        reference,
        relations,
      );
    }
    if (rest.length > 0) {
      throw new Error(
        `Mapping for '${segment}' does not support further nesting in '${reference}'`,
      );
    }
    return { relations, mapping: toBaseMapperEntry(fieldEntry) };
  }

  // A segment the relation does not declare falls back to the table's own column of that name.
  const inferredColumn =
    segment in relation.table ? (relation.table as never)[segment] : undefined;

  if (inferredColumn !== undefined) {
    if (rest.length > 0) {
      throw new Error(
        `Unable to resolve nested path '${segment}.${rest.join(".")}' for relation '${reference}'`,
      );
    }
    return { relations, mapping: toBaseMapperEntry(inferredColumn) };
  }

  throw new Error(
    `No mapping found for relation segment '${segment}' in reference '${reference}'`,
  );
};

/**
 * Resolve a reference to its mapping. An exact entry wins; otherwise the longest mapped prefix
 * that is a relation resolves the rest of the path through that relation's fields.
 */
export const resolveFieldReference = (
  reference: string,
  mapper: Mapper,
): ResolvedMapping => {
  const direct = getMappingEntry(reference, mapper);
  if (direct !== undefined) {
    if (isScopedRelationEntry(direct)) {
      return direct.resolve();
    }
    if (isMappingConfig(direct) && direct.relation) {
      return resolveRelationField(direct.relation, [], reference, [], false);
    }
    return { relations: [], mapping: direct };
  }

  for (const [prefix, suffix] of referencePrefixes(reference)) {
    const entry = getMappingEntry(prefix, mapper);
    if (!entry || !isMappingConfig(entry) || !entry.relation) {
      continue;
    }
    return resolveRelationField(entry.relation, suffix, reference, []);
  }

  throw new Error(`No mapping found for reference: ${reference}`);
};

/** The relation chain a collection reference iterates, root first. */
const resolveRelationChain = (
  reference: string,
  mapper: Mapper,
): RelationMapping[] => {
  const direct = getMappingEntry(reference, mapper);
  if (direct !== undefined) {
    if (isScopedRelationEntry(direct)) {
      return direct.resolve().relations;
    }
    if (isMappingConfig(direct) && direct.relation) {
      return [direct.relation];
    }
  }

  for (const [prefix, suffix] of referencePrefixes(reference)) {
    const entry = getMappingEntry(prefix, mapper);
    if (!entry) {
      continue;
    }
    // A prefix inside an enclosing lambda scope carries the relations already joined to reach it.
    const { relations, mapping } = isScopedRelationEntry(entry)
      ? entry.resolve()
      : { relations: [], mapping: entry };
    if (isMappingConfig(mapping) && mapping.relation) {
      return resolveRelationField(mapping.relation, suffix, reference, relations)
        .relations;
    }
  }

  throw new Error(`No relation mapping found for reference: ${reference}`);
};

/** A reference that resolved to a whole relation stands for that relation's default `field`. */
export const resolveRelationDefaultField = (
  resolved: ResolvedMapping,
  reference: string,
): ResolvedMapping => {
  if (!isRelationValue(resolved.mapping)) {
    return resolved;
  }
  const defaultField = resolved.mapping.relation.field;
  if (!defaultField) {
    throw new Error(
      `Relation mapping for '${reference}' does not define a default field`,
    );
  }
  return {
    relations: resolved.relations,
    mapping: toBaseMapperEntry(defaultField),
  };
};

// -- lambda scopes -------------------------------------------------------------------------------

/**
 * The scope a collection macro's lambda body is translated in: a mapper that resolves the lambda
 * variable to an element of the collection, and the relation chain that collection lives behind,
 * split into the relation being iterated (`primaryRelation`) and the hops leading to it.
 */
export interface CollectionScope {
  mapper: Mapper;
  primaryRelation: RelationMapping;
  leadingRelations: RelationMapping[];
  /** Every relation of the chain: the lambda body's subqueries are already correlated to them. */
  skipRelations: Set<RelationMapping>;
}

export const createCollectionScope = (
  collectionReference: string,
  variableName: string,
  mapper: Mapper,
): CollectionScope => {
  const relationChain = resolveRelationChain(collectionReference, mapper);
  const primaryRelation = relationChain[relationChain.length - 1];
  if (!primaryRelation) {
    throw new Error(
      `No relation mapping found for reference: ${collectionReference}`,
    );
  }
  const leadingRelations = relationChain.slice(0, -1);

  const scopedMapper = (reference: string): MapperEntry | undefined => {
    if (reference === variableName) {
      const resolved = resolveRelationField(
        primaryRelation,
        [],
        collectionReference,
        leadingRelations,
      );
      // An element of a scalar collection is a VALUE in CEL, so a NULL element is an explicit
      // null rather than a missing attribute.
      if (isScalarCollection(getMappingEntry(collectionReference, mapper))) {
        if (isColumn(resolved.mapping)) {
          resolved.mapping = {
            column: resolved.mapping,
            nullAttributeRepresentation: "explicit",
          };
        } else if (isMappingConfig(resolved.mapping)) {
          resolved.mapping = {
            ...resolved.mapping,
            nullAttributeRepresentation: "explicit",
          };
        }
      }
      return makeScopedRelationEntry(resolved);
    }

    if (reference.startsWith(`${variableName}.`)) {
      const remainder = reference.slice(variableName.length + 1);
      return makeScopedRelationEntry(
        resolveRelationField(
          primaryRelation,
          remainder.split("."),
          `${collectionReference}.${remainder}`,
          leadingRelations,
        ),
      );
    }

    return getMappingEntry(reference, mapper);
  };

  return {
    mapper: scopedMapper,
    primaryRelation,
    leadingRelations,
    skipRelations: new Set([primaryRelation, ...leadingRelations]),
  };
};

// -- columns -------------------------------------------------------------------------------------

/** The drizzle column an operand names, if it names one directly. */
export const columnForOperand = (
  operand: PlanExpressionOperand,
  mapper: Mapper,
): AnyColumn | undefined => {
  if (!isNameOperand(operand)) {
    return undefined;
  }
  const { mapping } = resolveFieldReference(operand.name, mapper);
  if (isMappingConfig(mapping)) {
    return mapping.column;
  }
  return isColumn(mapping) ? mapping : undefined;
};

/** A mapping read as a scalar SQL value: only a column (bare or configured) qualifies. */
export const buildColumnExpression = (
  mapping: BaseMapperEntry,
  reference: string,
): SQL => {
  if (isRelationValue(mapping)) {
    throw new Error(
      `Cannot use relation '${reference}' as a scalar value expression`,
    );
  }
  if (typeof mapping === "function") {
    throw new Error(
      `Cannot use transform mapping for '${reference}' as a value expression`,
    );
  }
  if (isMappingConfig(mapping)) {
    if (mapping.relation) {
      throw new Error(
        `Cannot use relation mapping for '${reference}' as a scalar value expression`,
      );
    }
    if (!mapping.column) {
      throw new Error(
        `Mapping for '${reference}' requires a column to be used as a value expression`,
      );
    }
    return sql`${mapping.column}`;
  }
  if (!isColumn(mapping)) {
    throw new Error(`Expected column mapping for '${reference}'`);
  }
  return sql`${mapping}`;
};
